pci: Add a quirk for devices with broken FLR reset

Some devices advertise FLReset+ but do not support it. Add a quirk for
some devices to fallback to an alternate reset method.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>

---
 drivers/pci/quirks.c    | 12 ++++++++++++
 include/linux/pci_ids.h |  2 ++
 2 files changed, 14 insertions(+)

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index ec2911c..a82ebfd 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3387,6 +3387,18 @@ static void quirk_no_pm_reset(struct pci_dev *dev)
 DECLARE_PCI_FIXUP_CLASS_HEADER(PCI_VENDOR_ID_ATI, PCI_ANY_ID,
 			       PCI_CLASS_DISPLAY_VGA, 8, quirk_no_pm_reset);
 
+static void quirk_no_flr_reset(struct pci_dev *dev)
+{
+	dev->dev_flags |= PCI_DEV_FLAGS_NO_FLR_RESET;
+}
+
+DECLARE_PCI_FIXUP_CLASS_HEADER(PCI_VENDOR_ID_LSI_LOGIC,
+			       PCI_DEVICE_ID_LSI_SAS3008,
+			       PCI_CLASS_STORAGE_SAS, 8, quirk_no_flr_reset);
+DECLARE_PCI_FIXUP_CLASS_HEADER(PCI_VENDOR_ID_LSI_LOGIC,
+			       PCI_DEVICE_ID_LSI_MEGARAID_3108,
+			       PCI_CLASS_STORAGE_SAS, 8, quirk_no_flr_reset);
+
 /*
  * Thunderbolt controllers with broken MSI hotplug signaling:
  * Entire 1st generation (Light Ridge, Eagle Ridge, Light Peak) and part
diff --git a/include/linux/pci_ids.h b/include/linux/pci_ids.h
index ab20dc5..6c552cb 100644
--- a/include/linux/pci_ids.h
+++ b/include/linux/pci_ids.h
@@ -216,6 +216,8 @@
 #define PCI_DEVICE_ID_LSI_SAS1066E	0x005A
 #define PCI_DEVICE_ID_LSI_SAS1068E	0x0058
 #define PCI_DEVICE_ID_LSI_SAS1078	0x0060
+#define PCI_DEVICE_ID_LSI_SAS3008       0x0097
+#define PCI_DEVICE_ID_LSI_MEGARAID_3108 0x005d
 
 #define PCI_VENDOR_ID_ATI		0x1002
 /* Mach64 */
